### DISCUSSION TITLE
Add an `IFunc<T, TResult>` interface

### DISCUSSION THREAD

**Discussion by charlesroddie** [dotnet/csharplang#5323] [Language Ideas]
Proposal:

1. Create an interface:

```fsharp
type IFunc<'a,'b> =
    abstract Invoke(x:'a) : 'b
```

2. Have `Func<T, TResult>` implement `IFunc<T, TResult>`, and have `Func<T0, ... Tn, TResult>` implement `Func<ValueTuple<T0...Tn>, TResult>`.

Advantages:

- This can be a common cross-language function interface, which `FSharpFunc` can also implement.
- It allows future consolidation of the functionality of `Func` and `Action` if/when dotnet gets a [`unit` type](https://github.com/dotnet/csharplang/discussions/1604).
- It allows for [invokable objects](https://github.com/fsharp/fslang-suggestions/issues/1085#issuecomment-940282229) which may be useful syntactic sugar in ML.

**Comment by huoyaoyuan** [comment]
This is more F#-ish, not C#-ish. C# tends to be more tightly integrated with the runtime, and prefer performance more than too much expressiveness. This may fall in range of HKT.
